### DISCUSSION TITLE
Qustion of Iteration Safety

### DISCUSSION THREAD

**Discussion by Cirnoo** [simdjson/simdjson#2215] [General]
Hi,

I found the section on `Iteration Safety` in the documentation and tried the provided code. According to the documentation, c2 is unsafe, but in my tests, c2 behaves correctly while c1 unexpectedly changes its value. Could you please help me understand these results? Thank you very much.

MacOS m2 arm64

```
   using namespace simdjson;
    ondemand::parser parser;
    const padded_string json = R"({ "parent": {"child1": {"name": "John"} , "child2": {"name": "Daniel"}} })"_padded;
    auto doc = parser.iterate(json);
    ondemand::object parent = doc["parent"];
    // parent owns the focus
    ondemand::object c1 = parent["child1"];
    // c1 owns the focus
    //
    std::cout  << "C1="<< c1 << std::endl;
    std::string_view as = c1["name"];

    if (std::string_view(c1["name"]) != "John") {
        // ...
        std::string_view ss = c1["name"];
    }
    // c2 attempts to grab the focus from parent but fails
    ondemand::object c2 = parent["child2"];
    if (std::string_view(c2["name"]) != "Daniel") {
        return false;
    }
    std::cout << "C2=" << c2 << std::endl;
    std::cout << "C1=" << c1 << std::endl;
```
The output
```
C1={"name": "John"}
C2={"name": "Daniel"}
C1={"name": "John"} , "child2": {"name": "Daniel"}
```

**Comment by lemire** [comment]
I think you mean to use the following:

```cpp

  ondemand::parser parser;
  const padded_string json = R"({ "parent": {"child1": {"name": "John"} , "child2": {"name": "Daniel"}} })"_padded;
  auto doc = parser.iterate(json);
  ondemand::object parent = doc["parent"];
  // parent owns the focus
  ondemand::object c1 = parent["child1"];
  // c1 owns the focus
  //
  std::string_view as1 = c1["name"];
  // We have that as1 == "John", as long as 'parser' and 'json' live
  // c2 attempts to grab the focus from parent but fails
  ondemand::object c2 = parent["child2"];
  // c2 owns the focus, at this point c1 is invalid
  std::string_view as2 = c2["name"];
  // We have that as2 == "Daniel", as long as 'parser' and 'json' live
  std::cout << as1 << " " << as2 << std::endl; // prints John Daniel
```

>  According to the documentation, c2 is unsafe, but in my tests, c2 behaves correctly while c1 unexpectedly changes its value.

They are both safe while valid.

You can hold just one child at a time, so by the time you do `ondemand::object c2 = parent["child2"];`, then `ondemand::object c1 = parent["child1"]` is invalid and should not be used anymore. However, `c1` was valid for a time, and by the time you create `c2` it is then valid. The issue is to use both of them at the same time.

You need to consume `c1` before you move on to `c2`. Once you have access to `c2`, you no longer have access to `c1`.

If you compile your code in debug mode, which is highly recommended, you should get immediately the error...

```
simdjson::simdjson_error: OUT_OF_ORDER_ITERATION: Objects and arrays can only be iterated when they are first encountered.
```

This tells you that your code is unsafe.

**Reply by Cirnoo** [reply]
Thank you very much!
